feat(curriculum): add polygon area calculator lab (#59460)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-polygon-area-calculator/5e444147903586ffb414c94f.md

---
id: 5e444147903586ffb414c94f
title: Build a Polygon Area Calculator

> challengeType: 23

Reviewer: ```suggestion
challengeType: 27
```
We have a new challenge type https://github.com/freeCodeCamp/freeCodeCamp/pull/59732

dashedName: build-a-polygon-area-calculator
---

# --description--

In this project, you will use object-oriented programming to create a `Rectangle` class and a `Square` class. The `Square` class should be a subclass of `Rectangle` and inherit its methods and attributes.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.

**User Stories:**

1. You should create a `rectangle` class.
2. When a Rectangle object is created, it should be initialized with `width` and `height` attributes. The class should also contain the following methods:

    - `set_width`: Sets the width of the rectangle.
    - `set_height`: Sets the height of the rectangle.
    - `get_area`: Returns area (`width * height`).
    - `get_perimeter`: Returns perimeter (`2 * width + 2 * height`).

>     - `get_diagonal`: Returns diagonal (`(width ** 2 + height ** 2) ** .5`).

Reviewer 2: We should find a different way of providing formulaes to them. Because I can see a lot of campers copying this exactly and being confused on why it's wrong or forgetting that they need to include `self`. 

<img width="1512" alt="Screenshot 2025-04-20 at 3 01 19 PM" src="https://github.com/user-attachments/assets/6ecc425a-27f8-44b2-a13f-16a1abc2c975" />

Reviewer 3: could we use mathjax to give the formulas in mathematical notation?

Author: Can we get a few suggestions on this?
I wasn't able to add the mathjax expressions, they didn't render as expected.

Reviewer 2: After playing around with this, I realized that math jax is not supported in the FSD cert. So I'll make a PR for that. Once I was able to get Math jax working here is the inline result

<img width="525" alt="Screenshot 2025-05-01 at 1 03 40 PM" src="https://github.com/user-attachments/assets/791c8557-6af1-4489-bb85-95a1f86f574d" />

```md
1. When a `Rectangle` object is created, it should be initialized with `width` and `height` attributes. The class should also contain the following methods:

    - `set_width`: Sets the width of the rectangle.
    - `set_height`: Sets the height of the rectangle.
    - `get_area`: Returns area (\\( \text{width} \times \text{height} \\)).
    - `get_perimeter`: Returns perimeter \\( 2(\text{width} + \text{height}) \\).
    - `get_diagonal`: Returns diagonal \\( \sqrt{\text{width}^2 + \text{height}^2} \\).
```

What do you think?

Author: updated!

![image](https://github.com/user-attachments/assets/f468ce3c-e564-424a-a2b7-eee7c776067a)

>     - `get_picture`: Returns a string that represents the shape using lines of '\*'. The number of lines should be equal to the height and the number of '\*' in each line should be equal to the width. There should be a new line (`\n`) at the end of each line. If the width or height is larger than `50`, this should return the string: `'Too big for picture.'`.

Reviewer 2: Not sure if '*' this is supposed to be a string inside backticks or just inside backticks.
This is what it currently looks like for me

<img width="458" alt="Screenshot 2025-04-20 at 7 20 22 PM" src="https://github.com/user-attachments/assets/2ab37176-de68-44f8-a220-c8ad912b364a" />

>     - `get_picture`: Returns a string that represents the shape using lines of `*`. The number of lines should be equal to the height and the number of '\*' in each line should be equal to the width. There should be a new line (`\n`) at the end of each line. If the width or height is larger than `50`, this should return the string: `'Too big for picture.'`.

Reviewer 2: it looks like this should be in backticks too. otherwise it looks like this rendered on the screen.

<img width="447" alt="Screenshot 2025-05-06 at 5 01 23 PM" src="https://github.com/user-attachments/assets/3a1fcf3c-0507-4bcc-a836-4bbd6c4e0c9b" />

```suggestion
    - `get_picture`: Returns a string that represents the shape using lines of `*`. The number of lines should be equal to the height and the number of `*` in each line should be equal to the width. There should be a new line (`\n`) at the end of each line. If the width or height is larger than `50`, this should return the string: `'Too big for picture.'`.
```

    - `get_amount_inside`: Takes another shape (square or rectangle) as an argument. Returns the number of times the passed in shape could fit inside the shape (with no rotations). For instance, a rectangle with a width of `4` and a height of `8` could fit in two squares with sides of `4`.

3. If an instance of a `Rectangle` is represented as a string, it should look like: `'Rectangle(width=5, height=10)'`.

4. You should create a `Square` class that subclasses `Rectangle`.
5. When a `Square` object is created, it should be initialized with a single side length. The `__init__` method should store the side length in both the `width` and `height` attributes from the `Rectangle` class.

> 6. The square class should contain the following methods:

Reviewer 2: ```suggestion
6. The `Square` class should contain the following methods:
```

---
id: 5e444147903586ffb414c94f
title: Build a Polygon Area Calculator
challengeType: 27
dashedName: build-a-polygon-area-calculator
---

# --description--

In this project, you will use object-oriented programming to create a `Rectangle` class and a `Square` class. The `Square` class should be a subclass of `Rectangle` and inherit its methods and attributes.

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.

**User Stories:**

1. You should create a `Rectangle` class.

2. When a `Rectangle` object is created, it should be initialized with `width` and `height` attributes. The class should also contain the following methods:

    - `set_width`: Sets the width of the rectangle.
    - `set_height`: Sets the height of the rectangle.
    - `get_area`: Returns area (\\( \text{width} \times \text{height} \\)).
    - `get_perimeter`: Returns perimeter \\( 2(\text{width} + \text{height}) \\).
    - `get_diagonal`: Returns diagonal \\( \sqrt{\text{width}^2 + \text{height}^2} \\).

>     - `get_picture`: Returns a string that represents the shape using lines of `*`. The number of lines should be equal to the height and the number of `*` in each line should be equal to the width. There should be a new line (`\n`) at the end of each line. If the width or height is larger than `50`, this should return the string: `'Too big for picture.'`.

Reviewer: ```suggestion
    - `get_picture`: Returns a string that represents the shape using lines of `*`. The number of lines should be equal to the height and the number of `*` in each line should be equal to the width. There should be a new line (`\n`) at the end of each line. If the width or height is larger than `50`, this should return the string: `Too big for picture.`.
```

    - `get_amount_inside`: Takes another shape (square or rectangle) as an argument. Returns the number of times the passed in shape could fit inside the shape (with no rotations). For instance, a rectangle with a width of `4` and a height of `8` could fit in two squares with sides of `4`.

> 3. If an instance of a `Rectangle` is represented as a string, it should look like: `'Rectangle(width=5, height=10)'`.

Reviewer: ```suggestion
3. If an instance of a `Rectangle` is represented as a string, it should look like: `Rectangle(width=5, height=10)`.
```

4. You should create a `Square` class that subclasses `Rectangle`.
5. When a `Square` object is created, it should be initialized with a single side length. The `__init__` method should store the side length in both the `width` and `height` attributes from the `Rectangle` class.
6. The `Square` class should contain the following methods:

    - `set_width`: Overrides the `set_width` method from the `Rectangle` class. It should set the width and height to the side length.
    - `set_height`: Overrides the `set_height` method from the `Rectangle` class. It should set the width and height to the side length.
    - `set_side`: Sets the height and width of the square equal to the side length.

7. The `Square` class should be able to access the `Rectangle` class methods.

> 8. If an instance of a `Square` is represented as a string, it should look like: `'Square(side=9)'`.

Reviewer: ```suggestion
8. If an instance of a `Square` is represented as a string, it should look like: `Square(side=9)`.
```

reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = runPython(testCode);
    assert(out);
  }
})
```

A square object should be an instance of the `Square` class and the `Rectangle` class.

```js
({
  test: () => {
    pyodide.FS.writeFile('/home/pyodide/shape_calculator.py', code);
    pyodide.FS.writeFile('/home/pyodide/test_module.py',`
import unittest
import shape_calculator
from importlib import reload
reload(shape_calculator)
class UnitTests(unittest.TestCase):
    maxDiff = None
    def setUp(self):
        self.sq = shape_calculator.Square(5)
    
    def test_square_is_square_and_rectangle(self):
        actual = isinstance(self.sq, shape_calculator.Square) and isinstance(self.sq, shape_calculator.Rectangle)
        expected = True
        self.assertEqual(actual, expected, 'Expected square object to be an instance of the Square class and the Rectangle class.')    
`);
    const testCode = `
from unittest import main
from importlib import reload
import test_module
reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = runPython(testCode);
    assert(out);
  }
})
```

> The string representation of `Rectangle(3, 6)` should be `'Rectangle(width=3, height=6)'`.

Reviewer: ```suggestion
The string representation of `Rectangle(3, 6)` should be `Rectangle(width=3, height=6)`.
```

```js
({
  test: () => {
    pyodide.FS.writeFile('/home/pyodide/shape_calculator.py', code);
    pyodide.FS.writeFile('/home/pyodide/test_module.py',`
import unittest
import shape_calculator
from importlib import reload
reload(shape_calculator)
class UnitTests(unittest.TestCase):
    maxDiff = None
    def setUp(self):
        self.rect = shape_calculator.Rectangle(3, 6)
    
    def test_rectangle_string(self):
        actual = str(self.rect)
        expected = "Rectangle(width=3, height=6)"
        self.assertEqual(actual, expected, 'Expected string representation of rectangle to be "Rectangle(width=3, height=6)"')    
`);
    const testCode = `
from unittest import main
from importlib import reload
import test_module
reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = runPython(testCode);
    assert(out);
  }
})
```

> The string representation of `Square(5)` should be `'Square(side=5)'`.

Reviewer: ```suggestion
The string representation of `Square(5)` should be `Square(side=5)`.
```

from unittest import main
from importlib import reload
import test_module
reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = runPython(testCode);
    assert(out);
  }
})
```

The `.get_picture()` method should return a different string representation of a `Square` instance.

```js
({
  test: () => {
    pyodide.FS.writeFile('/home/pyodide/shape_calculator.py', code);
    pyodide.FS.writeFile('/home/pyodide/test_module.py',`
import unittest
import shape_calculator
from importlib import reload
reload(shape_calculator)
class UnitTests(unittest.TestCase):
    maxDiff = None
    def setUp(self):
        self.sq = shape_calculator.Square(5)
    
    def test_square_picture(self):
        self.sq.set_side(2)
        actual = self.sq.get_picture()
        expected = "**\\n**\\n"
        self.assertEqual(actual, expected, 'Expected square picture to be different.')    
`);
    const testCode = `
from unittest import main
from importlib import reload
import test_module
reload(test_module)
t = main(module='test_module', exit=False)
t.result.wasSuccessful()
`;
    const out = runPython(testCode);
    assert(out);
  }
})
```

> The `.get_picture()` method should return the string `'Too big for picture.'` if the `width` or `height` attributes are larger than `50`.

Reviewer: ```suggestion
The `.get_picture()` method should return the string `Too big for picture.` if the `width` or `height` attributes are larger than `50`.
```
